pkg/kapp/cmd/core/deps_factory.go

```diff
@@ -107,3 +117,14 @@ func (f *DepsFactoryImpl) summarizeNodes(config *rest.Config) string {
 		return fmt.Sprintf("%s, %d+", oldestNode.Name, len(nodes.Items)-1)
 	}
 }
+
+func (f *DepsFactoryImpl) newWarningHandler() rest.WarningHandler {
+	if !f.Warnings {
+		return rest.NoWarnings{}
+	}
+	options := rest.WarningWriterOptions{
+		Deduplicate: true,
+	}
+	warningWriter := rest.NewWarningWriter(os.Stderr, options)
```

Comment:
@cppforlife Redirecting the warnings to Stderr as that' what k8s does, should we redirect it to Stdout instead?

Reply:
should be using o.ui.PrintLinef as the destination.

Reply:
I had tried adding a custom warning handler so that we could use ui.PrintLinef, but didn't proceed with it because of the following 2 reasons:

- Code duplication: I almost wrote the exact same code as provided by [client-go](https://github.com/kubernetes/client-go/blob/2961e1de2c1350e8532945d8aca4a0b1082f2232/rest/warnings.go#L94), except that they used fmt.Printf() and I used ui.PrintLinef(). (I mentioned this [previously](https://github.com/vmware-tanzu/carvel-kapp/pull/266#discussion_r690256775) and I am really sorry for adding a new comment for an already existing discussion).
- New Lines: The warning messages provided by k8s have new line appended at the end and since ui.PrintLinef uses fmt.Println(), every warning had an extra line in between them (unless I removed the new lines).

However, I will add it again and then we can decide whether to move forward with it or not.

Reply:
We can also provide an intermediate buffer to client go warning writer and then provided the stringified version to ui.PrintLinef, althought I am not convinced with this approach as well.
Let me know your thoughts as I might have missed on some better alternatives :)

Reply:
NewWarningWriter takes an io.Writer. so we can add a little adapter

```
type uiWriter struct { ui UI }
func (w uiWriter) Write(data []byte) (int, error) { w.ui.BeginLinef("%s", data); return len(data), nil }
```

Reply:
Done. And updated the test to check Stdout instead of Stderr.